rfcs/2021-06-01-7694-improve-monitor-build-performance.md

```diff
@@ -0,0 +1,115 @@
+# RFC 7694 - 2021-06-01 - Improving and Monitoring Build Performance of Vector
+
+This RFC describes the need for improvement to not only the build performance of Vector, but also
+monitoring around the performance such that we can maintain healthy performance in the long term.
+
+## Scope
+
+- Build performance from the developer perspective and CI perspective
+- Utilization of our CI infrastructure
+- [RFC 7027: Vector core extraction](https://github.com/timberio/vector/issues/7027)
+
+## Motivation
+
+As time goes on, Vector has only gotten more expensive to build, both in terms of computational resources and time spent. In order to provide a best-in-class observability hub, we need a development cycle that affords us the ability to quickly test ideas and validate assumptions. When we spend too much time building Vector, that cycle becomes longer, and we lose agility and productivity.
+
+## Internal Proposal
+
+The plan of attack is multi-faceted, focusing both on improving the build performance, and laying
+the groundwork for understanding the performance over time in the future.
+
+### More granular compilation units
+
+@blt is already tackling this through his continued work on [RFC 7027](https://github.com/timberio/vector/issues/7027).  This RFC seeks to break up the project structure of Vector such that changes in a far-off part of the codebase don’t force recompilation of unrelated parts where possible.
+
+This will speed up the development phase, as developers are often working in narrow areas of the codebase.
+
+### Cached compilation results
+
+While the linker time typically dominates compiling Vector – due to its single-threaded nature – we can extract value by caching the compilation of crates where possible.  Crate-level caching is useful across the board: development, performance testing, and CI.
+
```

Comment:
This might be a dumb question, but is it possible to reduce linker time? For example, would eliminating dependencies help (e.g. through consolidation or elimination)?

Reply:
You inspired me to go out and do some more investigation and I found a promising avenue of using `lld` as an alternate linker which is significantly faster.  Going to add a new item for that.